packages/client/src/runtime/queryArgsTypecheck.ts

```diff
@@ -0,0 +1,177 @@
+  }
+}
+
+class UuidType implements Type {
+  acceptsValue(value: unknown): boolean {
+    return typeof value === 'string' && UUID_REGEX.test(value)
+  }
+}
+
+class IdType implements Type {
+  acceptsValue(value: unknown): boolean {
+    return typeof value === 'string'
+  }
+}
+
+class DateTimeType implements Type {
+  acceptsValue(value: unknown): boolean {
+    if (Object.prototype.toString.call(value) === '[object Date]') {
+      return true
+    }
+
+    if (typeof value !== 'string') {
+      return false
+    }
+
+    return RFC_3339_REGEX.test(value)
+  }
+}
+
+class StringType implements Type {
+  acceptsValue(value: unknown): boolean {
+    return typeof value === 'string'
+  }
+}
+
+class DecimalType implements Type {
+  acceptsValue(value: unknown): boolean {
+    if (Decimal.isDecimal(value) || typeof value === 'number') {
+      return true
+    }
+
+    if (isDecimalJsLike(value)) {
+      return true
+    }
+
+    if (typeof value === 'string') {
+      // https://github.com/MikeMcl/decimal.js/blob/master/decimal.js#L116
+      return /^\-?(\d+(\.\d*)?|\.\d+)(e[+-]?\d+)?$/i.test(value)
+    }
+
+    return false
+  }
+}
+
+class BytesType implements Type {
+  acceptsValue(value: unknown): boolean {
+    return Buffer.isBuffer(value)
+  }
+}
+
+class JsonType implements Type {
+  acceptsValue(value: unknown): boolean {
+    return true
+  }
+}
+
+class ListType implements Type {
+  readonly elementType: Type
+
+  constructor(elementType: Type) {
+    this.elementType = elementType
+  }
+
+  acceptsValue(value: unknown): boolean {
+    if (!Array.isArray(value)) {
+      return false
+    }
+
+    return value.every((item) => this.elementType.acceptsValue(item))
+  }
+}
+
+class EnumType implements Type {
+  readonly name: string
+  readonly values: string[]
+
+  constructor(name: string, choices: string[]) {
+    this.name = name
+    this.values = choices
+  }
+
+  acceptsValue(value: unknown): boolean {
+    return typeof value === 'string' && this.values.includes(value)
+  }
+}
+
+const dmmfTypesMapping = {
+  Null: NullType,
+  Boolean: BooleanType,
+  Int: IntType,
+  BigInt: BigIntType,
+  Float: FloatType,
+  UUID: UuidType,
+  ID: IdType,
+  DateTime: DateTimeType,
+  String: StringType,
+  Decimal: DecimalType,
+  Bytes: BytesType,
+  Json: JsonType,
+}
+
+export function makeScalarInputType(dmmfArgType: DMMF.SchemaArgInputType): Type {
+  let type: Type
+  const dmmfType = dmmfArgType.type
+  if (typeof dmmfType === 'string') {
+    type = typeByName(dmmfType)
+  } else if ((dmmfType as DMMF.SchemaEnum).values) {
+    type = new EnumType(dmmfType.name, (dmmfType as DMMF.SchemaEnum).values)
+  } else {
+    type = typeByName(dmmfType.name)
+  }
+
+  if (dmmfArgType.isList) {
+    return new ListType(type)
+  }
+
+  return type
+}
+
+function typeByName(name: string): Type {
+  const TypeConstructor = dmmfTypesMapping[name]
+  if (!TypeConstructor) {
+    throw new TypeError(`Unknown scalar type: ${name}`)
```

Comment:
Question: do we have a specific error type for the cases of "This should never have happened in real life, but it did and here we are"?